### PR TITLE
[HALX86] Disable Lazy IRQL in APIC HAL

### PR DESCRIPTION
**Pull request by DarkFire01**
## Purpose
-First couple steps to get ACPI APIC working.

## Proposed changes

_Describe what you propose to change/add/fix with this pull request._
  -Cleaned APIC header files 
  -Created files for building the respective HALs
  -Disabled LazyIRQL until a solution for it is found
         -APIC_LAZY_IRQL being enabled stops virtualbox from being able to boot at the current time, during my time attempting to solve the issue I discovered that it occurs after a raise IRQL event; this doesn't succeed and causes an asset fail before finishing the initialization of the later stages of the kernel.

### CODE REVIEW

File: hal/halx86/smp.cmake [hunk 1]
```diff
@@ -0,0 +1,12 @@
+list(APPEND HAL_SMP_SOURCE
```
Comment (Extravert-ir) [inline hal/halx86/smp.cmake:1]: This is empty now, I'd add this along with future changes

File: hal/halx86/include/apic.h [hunk 2]
```diff
@@ -1,253 +1,303 @@
...
+#define APIC_EAFR     0x0400 /* extended APIC Feature register (R/W) */
+#define APIC_EACR     0x0410 /* Extended APIC Control Register (R/W) */
+#define APIC_SEOI     0x0420 /* Specific End Of Interrupt Register (W) */
+#define APIC_EXT0LVTR 0x0500 /* Extended Interrupt 0 Local Vector Table */
+#define APIC_EXT1LVTR 0x0510 /* Extended Interrupt 1 Local Vector Table */
+#define APIC_EXT2LVTR 0x0520 /* Extended Interrupt 2 Local Vector Table */
+#define APIC_EXT3LVTR 0x0530 /* Extended Interrupt 3 Local Vector Table */
+
+#define MSR_APIC_BASE 0x0000001B
+#define IOAPIC_PHYS_BASE 0xFEC00000
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:90]: Is it the same for both i686 and amd64?

**Review by Extravert-ir — CHANGES_REQUESTED** [review]
Congratulations with the first PR!

**Comment by Extravert-ir** [comment]
BTW, it makes sense to do it in two commits - one for `apic.h` changes, another one for lazy IRQL disable

File: hal/halx86/mp/apicmp.h [hunk 3]
```diff
@@ -0,0 +1,253 @@
+/*
+ *
+ */
+
+#pragma once
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply hal/halx86/mp/apicmp.h:5]: This was the Apic.h inside the include folder before, It's the one that is required for building halmp.dll. It was just advised i keep it!

File: hal/halx86/apic/halinit.c [hunk 4]
```diff
@@ -28,15 +29,12 @@ HalpInitProcessor(
     IN ULONG ProcessorNumber,
     IN PLOADER_PARAMETER_BLOCK LoaderBlock)
 {
+
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply hal/halx86/apic/halinit.c:32]: The extra spaces were an accident while pushing, i didn't realize i accidentally sent those.

File: hal/halx86/apic/halinit.c [hunk 5]
```diff
@@ -28,15 +29,12 @@ HalpInitProcessor(
...
     /* Initialize the local APIC for this cpu */
     ApicInitializeLocalApic(ProcessorNumber);
 
     /* Initialize profiling data (but don't start it) */
     HalInitializeProfiling();
-
-    /* Initialize the timer */
-    //ApicInitializeTimer(ProcessorNumber);
```
[the remark this answers is not in the document]
Comment (DarkFire01) [reply hal/halx86/apic/halinit.c:38]: Not on purpose, I had rewritten that function on a local branch and accidentally committed some of the changes it looks like. That is where a few of these accident occured.

File: hal/halx86/mp/apicmp.h [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+/* Prototypes */
+VOID APICSendIPI(ULONG Target, ULONG Mode);
+VOID APICSetup(VOID);
+VOID HaliInitBSP(VOID);
+VOID APICSyncArbIDs(VOID);
+VOID APICCalibrateTimer(ULONG CPU);
+VOID HaliStartApplicationProcessor(ULONG Cpu, ULONG Stack);
+
+static __inline ULONG _APICRead(ULONG Offset)
```
Comment (GeoB99) [inline hal/halx86/mp/apicmp.h:184]: You might as well use `FORCEINLINE` for all of these routines in the header.

**Review by GeoB99 — COMMENTED** [review]
Nothing much to review from my part, other than something worthwhile to consider.

File: hal/halx86/apic/tsc.h [hunk 7]
```diff
@@ -19,4 +19,4 @@ VOID NTAPI HalpInitializeTsc(void);
 
 #endif
 
-#endif /* !_TSC_H_ */
+#endif /* !_TSC_H_ */
```
Comment (SergeGautherie) [inline hal/halx86/apic/tsc.h:22]: Revert this change.

File: hal/halx86/apic/tsc.c [hunk 8]
```diff
@@ -11,7 +11,6 @@
 #include <hal.h>
 #define NDEBUG
 #include <debug.h>
-
 #include "tsc.h"
```
Comment (SergeGautherie) [inline hal/halx86/apic/tsc.c:14]: Move this header up too?

File: hal/halx86/CMakeLists.txt [hunk 9]
```diff
@@ -60,15 +60,10 @@ if(ARCH STREQUAL "i386")
     # hal
     add_hal(hal SOURCES pic/halpic.rc COMPONENTS generic legacy up pic)
     add_hal(halacpi SOURCES acpi/halacpi.rc COMPONENTS generic acpi up pic)
-    add_hal(halaacpi SOURCES acpi/halacpi.rc COMPONENTS generic acpi up apic)
+    add_hal(halaacpi SOURCES apic/halaacpi.rc COMPONENTS generic acpi up apic)
     add_hal(halxbox SOURCES xbox/halxbox.rc COMPONENTS xbox up)
     add_hal(halpc98 SOURCES pc98/halpc98.rc COMPONENTS pc98 up)
 
-    #add_hal(halmps SOURCES up/halup.rc COMPONENTS generic legacy smp pic)
-    #add_hal(halmacpi SOURCES acpi/halacpi.rc COMPONENTS generic acpi smp pic)
```
Comment (SergeGautherie) [inline hal/halx86/CMakeLists.txt:68]: Why remove `halmacpi`?

File: hal/halx86/smp/amd64/halmacpi.rc [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#define REACTOS_VERSION_DLL
+#define REACTOS_STR_FILE_DESCRIPTION  "X64 Multiprocessor Hardware Abstraction Layer"
+#define REACTOS_STR_INTERNAL_NAME     "halmacpi"
```
Comment (SergeGautherie) [inline hal/halx86/smp/amd64/halmacpi.rc:3]: Do we actually need `smp` directory, in addition to the `mp` one?

File: hal/halx86/smp/amd64/halmacpi.rc [hunk 11]
```diff
@@ -0,0 +1,5 @@
+#define REACTOS_VERSION_DLL
+#define REACTOS_STR_FILE_DESCRIPTION  "X64 Multiprocessor Hardware Abstraction Layer"
```
Comment (SergeGautherie) [inline hal/halx86/smp/amd64/halmacpi.rc:2]: Could both files be merged, with an `#if ...`?
Where is this file(s) used, anyway?

File: hal/halx86/smp/amd64/halmacpi.rc (hunk 10 above)
Comment (DarkFire01) [reply hal/halx86/smp/amd64/halmacpi.rc:3]: The mp folder isn't going to be there forever, I was advised just to keep it for now. The code inside the MP folder is all meant for the older MP tables and I just decided to keep unchanged with the exception of me moving the APIC header file that was meant for them into that folder. I'm going to be focusing my time on halmacpi for now instead of halmp.

File: hal/halx86/smp/amd64/halmacpi.rc (hunk 11 above)
Comment (DarkFire01) [reply hal/halx86/smp/amd64/halmacpi.rc:2]: I'm working on both 32bit/64bit halmacpi at the same time. I will however change this into being one file as that does make much more sense! Thank you

File: hal/halx86/apic/halinit.c [hunk 12]
```diff
@@ -1,16 +1,16 @@
 /*
  * PROJECT:     ReactOS Hardware Abstraction Layer
  * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
- * PURPOSE:     Initialize the x86 HAL
- * COPYRIGHT:   Copyright 2011 Timo Kreuzer <timo.kreuzer@reactos.org>
+ * PURPOSE:     Header file for APIC hal
+ * COPYRIGHT:   2011 Timo Kreuzer <timo.kreuzer@reactos.org>
```
Comment (Extravert-ir) [inline hal/halx86/apic/halinit.c:5]:
```suggestion
 * COPYRIGHT:   Copyright 2011 Timo Kreuzer <timo.kreuzer@reactos.org>
```

File: hal/halx86/apic/tsc.h [hunk 13]
```diff
@@ -18,5 +18,4 @@ VOID NTAPI HalpInitializeTsc(void);
 #endif
 
 #endif
-
-#endif /* !_TSC_H_ */
+#endif /* !_TSC_H_ */
```
Comment (Extravert-ir) [inline hal/halx86/apic/tsc.h:21]: Please revert this file :)

File: hal/halx86/include/apic.h [hunk 14]
```diff
@@ -1,253 +1,301 @@
 /*
- *
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Header file for APIC hal
+ * COPYRIGHT:   2011 Timo Kreuzer <timo.kreuzer@reactos.org>
+ *              2021 Justin Miller <justinmiller100@gmail.com>
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:6]:
```suggestion
 * COPYRIGHT:   Copyright 2011 Timo Kreuzer <timo.kreuzer@reactos.org>
 *              Copyright 2021 Justin Miller <justinmiller100@gmail.com>
```

File: hal/halx86/include/apic.h [hunk 15]
```diff
@@ -1,253 +1,301 @@
...
-#define ERROR_VECTOR		    0xFE
-#define SPURIOUS_VECTOR		    0xFF  /* Must be 0xXF */
-
-/* CPU flags */
-#define CPU_USABLE   0x01  /* 1 if the CPU is usable (ie. can be used) */
-#define CPU_ENABLED  0x02  /* 1 if the CPU is enabled */
-#define CPU_BSP      0x04  /* 1 if the CPU is the bootstrap processor */
-#define CPU_TSC      0x08  /* 1 if the CPU has a time stamp counter */
-
-typedef struct _CPU_INFO
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:155]: You think this structure will not be useful?

File: hal/halx86/include/apic.h [hunk 16]
```diff
@@ -1,253 +1,301 @@
...
+#define APIC_EXT1LVTR 0x0510 /* Extended Interrupt 1 Local Vector Table */
+#define APIC_EXT2LVTR 0x0520 /* Extended Interrupt 2 Local Vector Table */
+#define APIC_EXT3LVTR 0x0530 /* Extended Interrupt 3 Local Vector Table */
+
+#define MSR_APIC_BASE 0x0000001B
+#define IOAPIC_PHYS_BASE 0xFEC00000
+#define APIC_CLOCK_INDEX 8
+#define ApicLogicalId(Cpu) ((UCHAR)(1<< Cpu))
+
+enum
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:93]: enums need a comment about where do they belong to

File: hal/halx86/include/apic.h [hunk 17]
```diff
@@ -1,253 +1,301 @@
...
+   Writing a value of 01h forces the NMI and 8259 INTR signals to pass through the APIC.
+*/
+#define IMCR_ADDRESS_PORT  (PUCHAR)0x0022
+#define IMCR_DATA_PORT     (PUCHAR)0x0023
+#define IMCR_SELECT        0x70
+#define IMCR_PIC_DIRECT    0x00
+#define IMCR_PIC_VIA_APIC  0x01
+
+#define ZERO_VECTOR 0x00 //IRQL 00
+#define APC_VECTOR           0x3D // IRQL 01
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:42]:
```suggestion
#define ZERO_VECTOR          0x00 // IRQL 00
#define APC_VECTOR           0x3D // IRQL 01
```

File: hal/halx86/include/apic.h [hunk 18]
```diff
@@ -1,253 +1,301 @@
...
-
-#define APIC_INTEGRATED(version) (version & 0xF0)
-
-typedef enum {
-  amPIC = 0,    /* IMCR and PIC compatibility mode */
-  amVWIRE       /* Virtual Wire compatibility mode */
-} APIC_MODE;
-
-#ifdef CONFIG_SMP
-#define MAX_CPU   32
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:133]: This definition will probably be needed in future

File: hal/halx86/include/apic.h [hunk 19]
```diff
@@ -1,253 +1,301 @@
...
+    struct
+    {
+        ULONG Vector:8;
+        ULONG SoftwareEnable:1;
+        ULONG FocusCPUCoreChecking:1;
+        ULONG ReservedMBZ:22;
+    };
+} APIC_SPURIOUS_INERRUPT_REGISTER;
+
+typedef union
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:170]:
```suggestion
typedef union _APIC_VERSION_REGISTER
```
Here and down the file

File: hal/halx86/include/apic.h [hunk 20]
```diff
@@ -1,253 +1,301 @@
...
+        ULONG Version:1;
+        ULONG SEOIEnable:1;
+        ULONG ExtApicIdEnable:1;
+        ULONG ReservedMBZ:29;
+    };
+} APIC_EXTENDED_CONTROL_REGISTER;
+
+typedef union _APIC_COMMAND_REGISTER
+{
+    ULONGLONG LongLong;
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:197]: For hardware stuff, I'd like the definitions to use types with bit count, this makes things more explicit. `UINT8`, `UINT16`, `UINT32`, `UINT64`
By the way, should we make them packed?

File: hal/halx86/include/apic.h [hunk 21]
```diff
@@ -1,253 +1,301 @@
...
+
+VOID
+NTAPI
+HalInitializeProfiling(VOID);
+
+VOID
+NTAPI
+HalpInitApicInfo(IN PLOADER_PARAMETER_BLOCK KeLoaderBlock); 
+
+VOID __cdecl ApicSpuriousService(VOID);
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:301]: Each file should have an extra line break in the end

File: hal/halx86/mp/apicmp.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Extravert-ir) [reply hal/halx86/mp/apicmp.h:5]: No need to keep it. If it has some useful definitions, they should be put into apic.h

**Review by Extravert-ir — CHANGES_REQUESTED** [review]

File: hal/halx86/include/apic.h (hunk 15 above)
Comment (DarkFire01) [reply hal/halx86/include/apic.h:155]: This won't end up being used.

File: hal/halx86/include/apic.h (hunk 18 above)
Comment (DarkFire01) [reply hal/halx86/include/apic.h:133]: It's done via the MADT tables, so it being defined in the header file will end up redundant.

File: hal/halx86/apic/tsc.h (hunk 13 above)
Comment (Extravert-ir) [reply hal/halx86/apic/tsc.h:22]: Add a new line to the end of file

File: hal/halx86/include/apic.h (hunk 17 above)
Comment (Extravert-ir) [reply hal/halx86/include/apic.h:42]: This one is not resolved

File: hal/halx86/include/apic.h [hunk 22]
```diff
@@ -1,253 +1,321 @@
...
+    {
+        UINT64 Reserved1:8;
+        UINT64 BootStrapCPUCore:1;
+        UINT64 Reserved2:2;
+        UINT64 Enable:1;
+        UINT64 BaseAddress:40;
+        UINT64 ReservedMBZ:12;
+    };
+} APIC_BASE_ADRESS_REGISTER;
+#include <poppack.h>
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:163]: No need to wrap every structure, you can do it once for all structures

File: hal/halx86/include/apic.h [hunk 23]
```diff
@@ -1,253 +1,311 @@
...
+    TIMER_DV_DivideBy32 = 8,
+    TIMER_DV_DivideBy64 = 9,
+    TIMER_DV_DivideBy128 = 10,
+    TIMER_DV_DivideBy1 = 11,
+};
+
+#include <pshpack1.h>
+typedef union _APIC_BASE_ADRESS_REGISTER
+{
+    UINT64 Long;
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:152]:
```suggestion
    UINT64 LongLong;
```

File: hal/halx86/include/apic.h [hunk 24]
```diff
@@ -1,253 +1,311 @@
...
+        UINT32 DeliveryStatus:1;
+        UINT32 Reserved1MBZ:1;
+        UINT32 RemoteIRR:1;
+        UINT32 TriggerMode:1;
+        UINT32 Mask:1;
+        UINT32 TimerMode:1;
+        UINT32 Reserved2MBZ:13;
+    };
+} LVT_REGISTER;
+#include <poppack.h>
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:242]:
```suggestion
```

File: hal/halx86/include/apic.h [hunk 25]
```diff
@@ -1,253 +1,311 @@
...
+/* IOAPIC Constants */
+enum
+{
+    IOAPIC_ID  = 0x00,
+    IOAPIC_VER = 0x01,
+    IOAPIC_ARB = 0x02,
+    IOAPIC_REDTBL = 0x10
+};
+
+#include <pshpack1.h>
```
Comment (Extravert-ir) [inline hal/halx86/include/apic.h:260]:
```suggestion
```

**Review by Extravert-ir — CHANGES_REQUESTED** [review]
